### PR TITLE
Modifications to EMOS related to using all realizations as the predictor

### PR DESCRIPTION
**Pull request by gavinevans**
Addresses Related to #854 

Description
This PR includes:
- Modifications to support the specification of the maximum number of iterations at the CLI. This is to support the usage of the realization option for the predictor of mean argument, where 200 iterations might not be sufficient to produce a stable solution, particularly if there are a large number of ensemble members.
- Make a correction to the initial guess when the realization option is chosen. In this case, the b coefficients that are solved for can essentially be interpreted as weights for each of the realizations, so therefore, it makes sense for each realization to start off with a weight of 1./no_of_realizations. This initial guess gives the minimisation a better chance to find a stable solution.
- Correct the datetime used in the CLI tests added for the estimate-emos-coefficients CLI.
- A further PR will add the CLI desired in #854.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 1]
```diff
@@ -246,15 +278,15 @@ def test_temperature_realizations_data_check(self):
             )
         else:
             predictor_data = np.array(
-                [[230.53659896, 241.80363361, 253.07066826],
-                 [264.33770292, 275.60473757, 286.87177222],
-                 [298.13880687, 309.40584153, 320.67287618]],
+                [[231.46936, 242.74925, 254.02914],
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: These changes are due to the change to the initial guess?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 2]
```diff
@@ -246,15 +278,15 @@ def test_temperature_realizations_data_check(self):
...
+                 [265.30902, 276.5889, 287.8688],
+                 [299.14868, 310.42856, 321.70844]],
                 dtype=np.float32
             )
             variance_data = np.array(
-                [[18.04589231, 18.04589231, 18.04589231],
-                 [18.04589231, 18.04589231, 18.04589231],
-                 [18.04589231, 18.04589231, 18.04589231]],
+                [[0.89248854, 0.89248854, 0.89248854],
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: The variance has changed a lot! Is this expected?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 3]
```diff
@@ -708,8 +744,7 @@ def test_coefficient_values_for_gaussian_distribution_realizations(self):
         if statsmodels_found:
             data = [-0.00114, -0.00006, 1.00037, -0.00196, 0.99999, -0.00315]
         else:
-            data = [4.30804737e-02, 1.39042785e+00, 8.99047025e-04,
-                    2.02661310e-01, 9.27197381e-01, 3.17407626e-01]
+            data = [0.9287, -0.0567, -0.00926, 0.16589, 0.74932, 0.64316]
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: These numbers have also changed a lot, can we have a chat about what they mean?

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 4]
```diff
@@ -87,6 +89,15 @@ def main(argv=None):
                              'ensemble mean ("mean") and the ensemble '
                              'realizations ("realizations") are supported as '
                              'options. Default: "mean".')
+    parser.add_argument('--max_iterations', metavar='MAX_ITERATIONS',
+                        type=np.int32, default=200,
+                        help='The maximum number of iterations allowed '
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: It might be worth saying in this docstring that the number of iterations should be increased if you are using realizations as a predictor.

File: tests/improver-estimate-emos-coefficients/02-gaussian.bats [hunk 5]
```diff
@@ -36,7 +36,7 @@
   KGO="estimate-emos-coefficients/gaussian/kgo.nc"
 
   # Estimate the EMOS coefficients and check that they match the kgo.
-  run improver estimate-emos-coefficients 'gaussian' '20170605T1500Z' \
+  run improver estimate-emos-coefficients 'gaussian' '20170605T0300Z' \
```
Comment (fionaRust) [inline tests/improver-estimate-emos-coefficients/02-gaussian.bats]: Why are these changes needed?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py (hunk 2 above)
Comment (bayliffe) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: Running the master code, again without statsmodel to access these tests, with 1000 iterations gives the following variances:
```
2.392e-10, 2.392e-10, 2.392e-10, 2.392e-10, 2.392e-10, 2.392e-10, 2.392e-10, 2.392e-10, 2.392e-10
```
which are tiny and not like either of the expected sets. Each increase in the number of iterations leads to a decrease in the variances as can be seen below, suggesting even the new expected variances given here are not the converged results, rather the values should be closer to 0:
```
No. iterations : variance
200 : 18.04589231
300 : 17.2682
400 : 16.7999
500 : 0.7757
600 : 0.0595
800 : 2.392e-10
```

If I run this PR version of this test with 1000 iterations I get variances of 7.9191e-13.

Removing the warning handling we get the expected warning about failure to converge:
```
The final iteration resulted in a percentage change that is greater than the accepted threshold of 5% i.e. [ 2.46329139 10.18034888  0.41063477  6.30490672  0.26428811  0.54476245]. 
A satisfactory minimisation has not been achieved. 
Last iteration: [ 0.92869924 -0.05670054 -0.00925947  0.16589068  0.74932268  0.6431581 ], 
Last-but-one iteration: [ 0.90637264 -0.05146157 -0.00922161  0.17705375  0.74734753  0.64668098]
Absolute difference: [0.0223266  0.00523897 0.00003787 0.01116307 0.00197515 0.00352288]
```

I guess the question is what do we want these tests to do? They show a consistent approach towards the solution with 200 iterations, but they do not converge. Is this sufficient, or do we want a converged solution? 800 iterations gives a converged solution.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 6]
```diff
@@ -190,11 +201,11 @@ def calculate_percentage_change_in_last_iteration(allvecs):
             args=(forecast_predictor_data, truth_data,
                   forecast_var_data, sqrt_pi, predictor_of_mean_flag),
             method="Nelder-Mead",
-            options={"maxiter": self.MAX_ITERATIONS, "return_all": True})
+            options={"maxiter": self.max_iterations, "return_all": True})
         if not optimised_coeffs.success:
             msg = ("Minimisation did not result in convergence after "
                    "{} iterations. \n{}".format(
-                       self.MAX_ITERATIONS, optimised_coeffs.message))
+                       self.max_iterations, optimised_coeffs.message))
```
Comment (bayliffe) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Sometimes we are getting nans in our initial guess and sometimes not, this can be seen by the intermittent appearance of the full warning shown below before convergence is reached.
```
UserWarning: Minimisation did not result in convergence after 680 iterations. 
Maximum number of iterations has been exceeded.
  warnings.warn(msg)
UserWarning: The final iteration resulted in a percentage change that is greater than the accepted threshold of 5% i.e. [292.69569882  61.02072573   0.0765011    0.02498964   0.00510256   0.00169396]. 
A satisfactory minimisation has not been achieved. 
Last iteration: [-0.00027948  0.00003825  0.00517252 -0.21116614  0.60728499  0.76590817], 
Last-but-one iteration: [ 0.00014504  0.00009813  0.00516856 -0.21111338  0.60731598  0.76589519]
Absolute difference: [0.00042452 0.00005988 0.00000395 0.00005276 0.00003099 0.00001297]
```
For some values of max_iterations we get only the following warning before convergence.
```
UserWarning: Minimisation did not result in convergence after 681 iterations. 
Maximum number of iterations has been exceeded.
```

I don't understand enough of what is going on to guess why some iterations are returning nans and some not.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 7]
```diff
@@ -333,13 +399,13 @@ def test_wind_speed_realizations_data_check(self):
                  [0.01406566, 0.01406566, 0.01406566]])
         else:
             predictor_data = np.array(
-                [[2.05799912, 9.73470204, 17.41140496],
-                 [25.08810788, 32.7648108, 40.44151372],
-                 [48.11821664, 55.79491955, 63.47162247]])
+                [[2.7344577, 10.298571, 17.862682],
+                 [25.426796, 32.99091, 40.55502],
+                 [48.119133, 55.683247, 63.24736]])
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: With 1000 iterations:
```
3. , 10.5, 18. , 25.5, 33. , 40.5, 48. , 55.5, 63.
```

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 8]
```diff
@@ -333,13 +399,13 @@ def test_wind_speed_realizations_data_check(self):
...
             variance_data = np.array(
-                [[4.26987243, 4.26987243, 4.26987243],
-                 [4.26987243, 4.26987243, 4.26987243],
-                 [4.26987243, 4.26987243, 4.26987243]])
+                [[0.03198197, 0.03198197, 0.03198197],
+                 [0.03198197, 0.03198197, 0.03198197],
+                 [0.03198197, 0.03198197, 0.03198197]])
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: With 1000 iterations:
```
8.142177e-10, 8.142177e-10, 8.142177e-10, 8.142177e-10, 8.142177e-10, 8.142177e-10, 8.142177e-10, 8.142177e-10, 8.142177e-10
```

You may have to explain to me why this minimiser makes everything 0 with enough iterations.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 3 above)
Comment (bayliffe) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: With 1000 iterations:
```
-7.21049e-07, 1.70712e-07, 0.00516975, -0.21115, 0.607316, 0.765887
```

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 9]
```diff
@@ -747,8 +782,8 @@ def test_coefficient_values_truncated_gaussian_distribution_realization(
             data = [0.11821805, -0.00474737, 0.17631301, 0.17178835,
                     0.66749225, 0.72287342]
         else:
-            data = [2.05550997, 0.10577237, 0.00028531,
-                    0.53208837, 0.67233013, 0.53704241]
+            data = [-0.057994, -0.084585, -0.001147,
+                    0.406686,  0.564365,  0.724325]
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: With 1000 iterations:
```
8.489104e-07, 1.426094e-05, 0.008252186, 0.2297337, 0.6313243, 0.7407101
```

File: tests/improver-estimate-emos-coefficients/01-help.bats [hunk 10]
```diff
@@ -73,6 +74,13 @@ optional arguments:
                         forecast mean. Currently the ensemble mean ("mean")
                         and the ensemble realizations ("realizations") are
                         supported as options. Default: "mean".
+  --max_iterations MAX_ITERATIONS
+                        The maximum number of iterations allowed until the
+                        minimisation has converged to a stable solution. If
+                        the maximum of iterations is reached, but then the
```
Comment (bayliffe) [inline tests/improver-estimate-emos-coefficients/01-help.bats]: Change in line with CLI is change accepted.

File: tests/improver-estimate-emos-coefficients/02-gaussian.bats (hunk 5 above)
Comment (gavinevans) [reply tests/improver-estimate-emos-coefficients/02-gaussian.bats]: This was my mistake when I set up the tests. This time should be the cycletime, see the description in the estimate-emos-coefficients, however, I mistakenly specified the validity time here. I've corrected that here.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 11]
```diff
@@ -87,6 +89,15 @@ def main(argv=None):
...
+                             'until the minimisation has converged to a '
+                             'stable solution. If the maximum of iterations '
+                             'is reached, but then the minimisation has '
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: Done.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 12]
```diff
@@ -79,21 +75,36 @@ class ContinuousRankedProbabilityScoreMinimisers(object):
...
+    def __init__(self, max_iterations=200):
+        """
+        Initialise class for performing minimisation of the Continuous
+        Ranked Probability Score (CRPS).
+
+        Kwargs:
+            max_iterations (int):
+                The maximum number of iterations allowed until the
+                minimisation has converged to a stable solution. If the
+                maximum of iterations is reached, but then the minimisation
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Done.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 13]
```diff
@@ -904,22 +926,30 @@ def __init__(self, calibration_method, distribution, desired_units=None,
                 String to specify the input to calculate the calibrated mean.
                 Currently the ensemble mean ("mean") and the ensemble
                 realizations ("realizations") are supported as the predictors.
+            max_iterations (int):
+                The maximum number of iterations allowed until the
+                minimisation has converged to a stable solution. If the
+                maximum of iterations is reached, but then the minimisation
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Done.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 14]
```diff
@@ -360,6 +371,12 @@ def __init__(self, distribution, current_cycle, desired_units=None,
...
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: Done.

**Comment by gavinevans** [comment]
I've now increased the default number of iterations to 1000 and refined the choice of initial guess, so that hopefully the solution will minimise sooner. Most of the code changes in the most recent commit are related to refactoring of the unit tests to try to make them a bit clearer by hopefully providing a simpler example, so that the coefficients generated can be more easily understood. Also hopefully the usage of statsmodels can be more easily understood in the unit tests, so I've added some tests that will be skipped depending upon whether statsmodels is in the environment.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 15]
```diff
@@ -45,27 +46,110 @@
...
                     "invalid escape sequence",
                     "can't resolve package from",
-                    "Collapsing a non-contiguous coordinate.",
                     "Minimisation did not result in"
                     " convergence",
                     "\nThe final iteration resulted in a percentage "
                     "change that is greater than the"
-                    " accepted threshold "]
+                    " accepted threshold ",
+                    "divide by zero encountered in true_divide",
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: Where do these new warnings come from?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py [hunk 16]
```diff
@@ -314,82 +626,104 @@ def test_wind_speed_realizations_data_check(self):
...
         """
-        calibration_method = "nonhomogeneous gaussian regression"
         distribution = "gaussian"
         desired_units = "degreesC"
-        plugin = Plugin(calibration_method, distribution, desired_units)
+        predictor_of_mean_flag = "realizations"
+        plugin = Plugin(
+            self.calibration_method, distribution, desired_units,
+            predictor_of_mean_flag=predictor_of_mean_flag,
+            max_iterations=10000)
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: Do you need this to be this high when statsmodels is available?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 17]
```diff
@@ -116,12 +120,10 @@ def test_statsmodels_mean(self, warning_list=None):
...
         historic_forecasts.concatenate_cube()
 
-        distribution = "gaussian"
-        desired_units = "degreesC"
         predictor_of_mean_flag = "mean"
 
-        if not statsmodels_found:
-            Plugin(distribution, desired_units,
+        if not STATSMODELS_FOUND:
+            Plugin(self.distribution, self.desired_units,
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: Which warning is this catching? And should this be a skip thing as well (and should some of the later tests be skipped too)?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py (hunk 15 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: The `divide by zero encountered in true_divide` is coming from the CRPS minimisation in the line: `if not np.isfinite(np.min(mu/sigma)):` implying that `sigma` is zero in this case. This is ok in this case, as the minimisation will just try again will slightly different coefficients, if a particular set of coefficients has resulted in a `sigma` of zero. 

The `invalid value encountered for` is occurs in multiple places, but I think that this is again related to the minimisation with runtime warnings of invalid value encountered in greater, invalid value encountered in less, invalid value encountered in true_divide, invalid value encountered in reduce, etc, which are arising from the minimisation.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py (hunk 16 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: I've reduced the number of max iterations when using statsmodels.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py (hunk 15 above)
Comment (fionaRust) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EnsembleCalibration.py]: Ok, I'm happy for these warnings to be ingored

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 4 above)
Comment (fionaRust) [reply lib/improver/cli/estimate_emos_coefficients.py]: Still outstanding @gavinevans

Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've now extended this docstring, @fionaRust.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 18]
```diff
@@ -754,16 +757,18 @@ def test_coefficients_gaussian_realizations_no_statsmodels(self):
         expected values, and the coefficient names also match
         expected values for a Gaussian distribution where the
         realizations are used as the predictor of the mean."""
-        data = np.array([0.0450917,  123.35383, -0.943704, -0.747887,
-                         -0.0180841, -0.655514], dtype=np.float32)
+        data = np.array([1.08333e+00, 1.09167e+00, 3.61111e-04,
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: Likewise, we might want to explain that we are not converging to the solution in these tests because the iterator is prone to escalating small errors that results in an overly sensitive test (e.g. to package versions, processor optimisation etc). You could stick the various comments in the doc-string at the top of the file if that is tidier as it will require a bit of explanation and happens in a few places.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 4 above)
Comment (fionaRust) [reply lib/improver/cli/estimate_emos_coefficients.py]: This should probably be copied into the other CLI too. Sorry

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 19]
```diff
@@ -518,23 +529,22 @@ def test_mean_predictor_estimate_coefficients(self):
         for the calibration coefficients, when the ensemble mean is used
         as the predictor. The coefficients are estimated using a linear model.
         """
-        data = np.array([1, 1, 2.66663, 1], dtype=np.float32)
+        data = np.array([0., 1., 1., 1.], dtype=np.float32)
```
Comment (cgsandford) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: This isn't a minor change.  What's going on here?

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 20]
```diff
@@ -567,7 +593,7 @@ def compute_initial_guess(
                             forecast_predictor.data.flatten()[
                                 combined_not_nan],
                             truth.data.flatten()[combined_not_nan]))
-                initial_guess = [1, 1, intercept, gradient]
+                initial_guess = [0, 1, intercept, gradient]
```
Comment (cgsandford) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: Why this change?  The others I can figure out the sense of, but not this one.
